chef/distro/common/man/man1/shef.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SHEF" "1" "November 2014" "Chef 10.34.6" "Chef Manual"
.
.SH "NAME"
\fBshef\fR \- Interactive Chef Console
.
.SH "SYNOPSIS"
\fBshef\fR [\fInamed configuration\fR] \fI(options)\fR
.
.TP
\fB\-S\fR, \fB\-\-server CHEF_SERVER_URL\fR
The chef server URL
.
.TP
\fB\-z\fR, \fB\-\-client\fR
chef\-client mode
.
.TP
\fB\-c\fR, \fB\-\-config CONFIG\fR
The configuration file to use
.
.TP
\fB\-j\fR, \fB\-\-json\-attributes JSON_ATTRIBS\fR
Load attributes from a JSON file or URL
.
.TP
\fB\-l\fR, \fB\-\-log\-level LOG_LEVEL\fR
Set the logging level
.
.TP
\fB\-s\fR, \fB\-\-solo\fR
chef\-solo shef session
.
.TP
\fB\-a\fR, \fB\-\-standalone\fR
standalone shef session
.
.TP
\fB\-v\fR, \fB\-\-version\fR
Show chef version
.
.TP
\fB\-h\fR, \fB\-\-help\fR
Show command options
.
.P
When no \-\-config option is specified, shef attempts to load a default configuration file:
.
.IP "\(bu" 4
If a \fInamed configuration\fR is given, shef will load ~/\.chef/\fInamed configuration\fR/shef\.rb
.
.IP "\(bu" 4
If no \fInamed configuration\fR is given shef will load ~/\.chef/shef\.rb if it exists
.
.IP "\(bu" 4
Shef falls back to loading /etc/chef/client\.rb or /etc/chef/solo\.rb if \-z or \-s options are given and no shef\.rb can be found\.
.
.IP "\(bu" 4
The \-\-config option takes precedence over implicit configuration paths\.
.
.IP "" 0
.
.SH "DESCRIPTION"
\fBshef\fR is an irb(1) (interactive ruby) session customized for Chef\. \fBshef\fR serves two primary functions: it provides a means to interact with a Chef Server interactively using a convenient DSL; it allows you to define and run Chef recipes interactively\.
.
.SH "SYNTAX"
Shef uses irb\'s subsession feature to provide multiple modes of interaction\. In addition to the primary mode which is entered on start, \fBrecipe\fR and \fBattributes\fR modes are available\.
.
.SH "PRIMARY MODE"
The following commands are available in the primary session:
.
.TP
\fBhelp\fR
Prints a list of available commands
.
.TP
\fBversion\fR
Prints the Chef version
.
.TP
\fBrecipe\fR
Switches to \fBrecipe\fR mode
.
.TP
\fBattributes\fR
Switches to \fBattributes\fR mode
.
.TP
\fBrun_chef\fR
Initiates a chef run
.
.TP
\fBreset\fR
reinitializes shef
.
.TP
\fBecho :on|:off\fR
Turns irb\'s echo function on or off\. Echo is \fIon\fR by default\.
.
.TP
\fBtracing :on|:off\fR
Turns irb\'s function tracing feature on or off\. Tracing is extremely verbose and expected to be of interest primarily to developers\.
.
.TP
\fBnode\fR
Returns the \fInode\fR object for the current host\. See knife\-node(1) for more information about nodes\.
.
.TP
\fBohai\fR
Prints the attributes of \fInode\fR
.
.P
In addition to these commands, shef provides a DSL for accessing data on the Chef Server\. When working with remote data in shef, you chain method calls in the form \fIobject type\fR\.\fIoperation\fR, where \fIobject type\fR is in plural form\. The following object types are available:
.
.IP "\(bu" 4
\fBnodes\fR
.
.IP "\(bu" 4
\fBroles\fR
.
.IP "\(bu" 4
\fBdata_bags\fR
.
.IP "\(bu" 4
\fBclients\fR
.
.IP "\(bu" 4
\fBcookbooks\fR
.
.IP "" 0
.
.P
For each \fIobject type\fR the following operations are available:
.
.TP
\fIobject type\fR\.all(\fI&block\fR)
Loads all items from the server\. If the optional code \fIblock\fR is given, each item will be passed to the block and the results returned, similar to ruby\'s \fBEnumerable#map\fR method\.
.
.TP
\fIobject type\fR\.show(\fIobject name\fR)
Aliased as \fIobject type\fR\.load
.
.IP
Loads the singular item identified by \fIobject name\fR\.
.
.TP
\fIobject type\fR\.search(\fIquery\fR, \fI&block\fR)
Aliased as \fIobject type\fR\.find
.
.IP
Runs a search against the server and returns the matching items\. If the optional code \fIblock\fR is given each item will be passed to the block and the results returned\.
.
.IP
The \fIquery\fR may be a Solr/Lucene format query given as a String, or a Hash of conditions\. If a Hash is given, the options will be ANDed together\. To join conditions with OR, use negative queries, or any advanced search syntax, you must provide give the query in String form\.
.
.TP
\fIobject type\fR\.transform(:all|\fIquery\fR, \fI&block\fR)
Aliased as \fIobject type\fR\.bulk_edit
.
.IP
Bulk edit objects by processing them with the (required) code \fIblock\fR\. You can edit all objects of the given type by passing the Symbol \fB:all\fR as the argument, or only a subset by passing a \fIquery\fR as the argument\. The \fIquery\fR is evaluated in the same way as with \fBsearch\fR\.
.
.IP
The return value of the code \fIblock\fR is used to alter the behavior of \fBtransform\fR\. If the value returned from the block is \fBnil\fR or \fBfalse\fR, the object will not be saved\. Otherwise, the object is saved after being passed to the block\. This behavior can be exploited to create a dry run to test a data transformation\.
.
.SH "RECIPE MODE"
Recipe mode implements Chef\'s recipe DSL\. Exhaustively documenting this DSL is outside the scope of this document\. See the following pages in the Chef documentation for more information:
.
.IP "\(bu" 4
\fIhttp://wiki\.opscode\.com/display/chef/Resources\fR
.
.IP "\(bu" 4
\fIhttp://wiki\.opscode\.com/display/chef/Recipes\fR
.
.IP "" 0
.
.P
Once you have defined resources in the recipe, you can trigger a convergence run via \fBrun_chef\fR
.
.SH "EXAMPLES"
.
.IP "\(bu" 4
A "Hello World" interactive recipe
.
.IP
chef > recipe
.
.br
chef:recipe > echo :off
.
.br
chef:recipe > file "/tmp/hello_world"
.
.br
chef:recipe > run_chef
.
.br
[Sat, 09 Apr 2011 08:56:56 \-0700] INFO: Processing file[/tmp/hello_world] action create ((irb#1) line 2)
.
.br
[Sat, 09 Apr 2011 08:56:56 \-0700] INFO: file[/tmp/hello_world] created file /tmp/hello_world
.
.br
chef:recipe > pp ls \'/tmp\'
.
.br
["\.",
.
.br
"\.\.",
.
.br
"hello_world"]
.
.IP "\(bu" 4
Search for \fInodes\fR by role, and print their IP addresses
.
.IP
chef > nodes\.find(:roles => \'monitoring\-server\') {|n| n[:ipaddress] }
.
.br
=> ["10\.254\.199\.5"]
.
.IP "\(bu" 4
Remove the role \fIobsolete\fR from every node in the system
.
.IP
chef > nodes\.transform(:all) {|n| n\.run_list\.delete(\'role[obsolete]\') }
.
.br
=> [node[chef098b2\.opschef\.com], node[ree\-woot], node[graphite\-dev], node[fluke\.localdomain], node[ghost\.local], node[kallistec]]
.
.IP "" 0
.
.SH "BUGS"
The name \fBshef\fR is clever in print but is confusing when spoken aloud\. Pronouncing \fBshef\fR as \fBchef console\fR is an imperfect workaround\.
.
.P
\fBshef\fR often does not perfectly replicate the context in which chef\-client(8) configures a host, which may lead to discrepancies in observed behavior\.
.
.P
\fBshef\fR has to duplicate much code from chef\-client\'s internal libraries and may become out of sync with the behavior of those libraries\.
.
.SH "SEE ALSO"
chef\-client(8) knife(1) \fIhttp://wiki\.opscode\.com/display/chef/Shef\fR
.
.SH "AUTHOR"
Chef was written by Adam Jacob \fIadam@opscode\.com\fR with many contributions from the community\. Shef was written by Daniel DeLeo\.
.
.SH "DOCUMENTATION"
This manual page was written by Daniel DeLeo \fIdan@opscode\.com\fR\. Permission is granted to copy, distribute and / or modify this document under the terms of the Apache 2\.0 License\.
.
.SH "CHEF"
Shef is distributed with Chef\. \fIhttp://wiki\.opscode\.com/display/chef/Home\fR


NEWS
Important Changes Since Chef 0.8
================================

  When upgrading from Chef 0.8.x to 0.9.x, be aware of a couple important changes.

    * 0.9 clients can read from 0.8 servers, but cannot save to 0.8 servers.
    * 0.8 clients cannot read or save from / to 0.9 servers.
    * The cookbook upload method is changed and cookbooks have to be
      uploaded to the server again.

  First, upgrade the server packages. Then upgrade the clients.

  Once the server is upgraded, cookbooks need to be uploaded again. Typically
  this is done from the Chef Repository (chef-repo) with knife. If the Chef
  Server is where the repository is located, simply:

      knife cookbook upload -a

  In the local chef-repo directory. If the repo is on another workstation,
  then upgrade Chef there and run the knife command, above.

  Two new configuration values are added to the server.rb:

    * sandbox_path
    * checksum_path

  Which are used for the new cookbook upload scheme.


chef/lib/chef/shef.rb
# Author:: Daniel DeLeo (<dan@kallistec.com>)
# Copyright:: Copyright (c) 2009 Daniel DeLeo
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'singleton'
require 'pp'
require 'etc'
require 'mixlib/cli'

require 'chef'
require 'chef/version'
require 'chef/client'
require 'chef/config'

require 'chef/shef/shef_session'
require 'chef/shef/ext'
require 'chef/json_compat'

# = Shef
# Shef is Chef in an IRB session. Shef can interact with a Chef server via the
# REST API, and run and debug recipes interactively.
module Shef
  LEADERS = Hash.new("")
  LEADERS[Chef::Recipe] = ":recipe"
  LEADERS[Chef::Node]   = ":attributes"

  class << self
    attr_accessor :client_type
    attr_accessor :options
    attr_accessor :env
    attr_writer   :editor
  end

  # Start the irb REPL with shef's customizations
  def self.start
    setup_logger
    # FUGLY HACK: irb gives us no other choice.
    irb_help = [:help, :irb_help, IRB::ExtendCommandBundle::NO_OVERRIDE]
    IRB::ExtendCommandBundle.instance_variable_get(:@ALIASES).delete(irb_help)

    parse_opts
    Chef::Config[:shell_config] = options.config

    # HACK: this duplicates the functions of IRB.start, but we have to do it
    # to get access to the main object before irb starts.
    ::IRB.setup(nil)

    irb = IRB::Irb.new

    init(irb.context.main)


    irb_conf[:IRB_RC].call(irb.context) if irb_conf[:IRB_RC]
    irb_conf[:MAIN_CONTEXT] = irb.context

    trap("SIGINT") do
      irb.signal_handle
    end

    catch(:IRB_EXIT) do
      irb.eval_input
    end
  end

  def self.setup_logger
    Chef::Config[:log_level] ||= :warn
    Chef::Log.init(STDERR)
    Mixlib::Authentication::Log.logger = Ohai::Log.logger = Chef::Log.logger
    Chef::Log.level = Chef::Config[:log_level] || :warn
  end

  # Shef assumes it's running whenever it is defined
  def self.running?
    true
  end

  # Set the irb_conf object to something other than IRB.conf
  # usful for testing.
  def self.irb_conf=(conf_hash)
    @irb_conf = conf_hash
  end

  def self.irb_conf
    @irb_conf || IRB.conf
  end

  def self.configure_irb
    irb_conf[:HISTORY_FILE] = "~/.shef_history"
    irb_conf[:SAVE_HISTORY] = 1000

    irb_conf[:IRB_RC] = lambda do |conf|
      m = conf.main

      conf.prompt_c       = "chef#{leader(m)} > "
      conf.return_format  = " => %s \n"
      conf.prompt_i       = "chef#{leader(m)} > "
      conf.prompt_n       = "chef#{leader(m)} ?> "
      conf.prompt_s       = "chef#{leader(m)}%l> "
    end
  end

  def self.leader(main_object)
    env_string = Shef.env ? " (#{Shef.env})" : ""
    LEADERS[main_object.class] + env_string
  end

  def self.session
    unless client_type.instance.node_built?
      puts "Session type: #{client_type.session_type}"
      client_type.instance.reset!
    end
    client_type.instance
  end

  def self.init(main)
    parse_json
    configure_irb

    session # trigger ohai run + session load

    session.node.consume_attributes(@json_attribs)

    Extensions.extend_context_object(main)

    main.version
    puts

    puts "run `help' for help, `exit' or ^D to quit."
    puts
    puts "Ohai2u#{greeting}!"
  end

  def self.greeting
    " #{Etc.getlogin}@#{Shef.session.node.fqdn}"
  rescue NameError, ArgumentError
    ""
  end

  def self.parse_json
    # HACK: copied verbatim from chef/application/client, because it's not
    # reusable as written there :(
    if Chef::Config[:json_attribs]
      begin
        json_io = open(Chef::Config[:json_attribs])
      rescue SocketError => error
        fatal!("I cannot connect to #{Chef::Config[:json_attribs]}", 2)
      rescue Errno::ENOENT => error
        fatal!("I cannot find #{Chef::Config[:json_attribs]}", 2)
      rescue Errno::EACCES => error
        fatal!("Permissions are incorrect on #{Chef::Config[:json_attribs]}. Please chmod a+r #{Chef::Config[:json_attribs]}", 2)
      rescue Exception => error
        fatal!("Got an unexpected error reading #{Chef::Config[:json_attribs]}: #{error.message}", 2)
      end

      begin
        @json_attribs = Chef::JSONCompat.from_json(json_io.read)
      rescue JSON::ParserError => error
        fatal!("Could not parse the provided JSON file (#{Chef::Config[:json_attribs]})!: " + error.message, 2)
      end
    end
  end

  def self.fatal!(message, exit_status)
    Chef::Log.fatal(message)
    exit exit_status
  end

  def self.client_type
    type = Shef::StandAloneSession
    type = Shef::SoloSession   if Chef::Config[:shef_solo]
    type = Shef::ClientSession if Chef::Config[:client]
    type = Shef::DoppelGangerSession if Chef::Config[:doppelganger]
    type
  end

  def self.parse_opts
    @options = Options.new
    @options.parse_opts
  end

  def self.editor
    @editor || Chef::Config[:editor] || ENV['EDITOR']
  end

  class Options
    include Mixlib::CLI

    def self.footer(text=nil)
      @footer = text if text
      @footer
    end

    banner("shef #{Chef::VERSION}\n\nUsage: shef [NAMED_CONF] (OPTIONS)")

    footer(<<-FOOTER)
When no CONFIG is specified, shef attempts to load a default configuration file:
* If a NAMED_CONF is given, shef will load ~/.chef/NAMED_CONF/shef.rb
* If no NAMED_CONF is given shef will load ~/.chef/shef.rb if it exists
* Shef falls back to loading /etc/chef/client.rb or /etc/chef/solo.rb if -z or
  -s options are given and no shef.rb can be found. 
FOOTER

    option :config_file,
      :short => "-c CONFIG",
      :long  => "--config CONFIG",
      :description => "The configuration file to use"

    option :help,
      :short        => "-h",
      :long         => "--help",
      :description  => "Show this message",
      :on           => :tail,
      :boolean      => true,
      :proc         => proc { print_help }

    option :log_level,
      :short  => "-l LOG_LEVEL",
      :long   => '--log-level LOG_LEVEL',
      :description => "Set the logging level",
      :proc         => proc { |level| Chef::Log.level = level.to_sym }

    option :standalone,
      :short        => "-a",
      :long         => "--standalone",
      :description  => "standalone shef session",
      :default      => true,
      :boolean      => true

    option :shef_solo,
      :short        => "-s",
      :long         => "--solo",
      :description  => "chef-solo shef session",
      :boolean      => true,
      :proc         => proc {Chef::Config[:solo] = true}

    option :client,
      :short        => "-z",
      :long         => "--client",
      :description  => "chef-client shef session",
      :boolean      => true

    option :json_attribs,
      :short => "-j JSON_ATTRIBS",
      :long => "--json-attributes JSON_ATTRIBS",
      :description => "Load attributes from a JSON file or URL",
      :proc => nil

    option :chef_server_url,
      :short => "-S CHEFSERVERURL",
      :long => "--server CHEFSERVERURL",
      :description => "The chef server URL",
      :proc => nil

    option :version,
      :short        => "-v",
      :long         => "--version",
      :description  => "Show chef version",
      :boolean      => true,
      :proc         => lambda {|v| puts "Chef: #{::Chef::VERSION}"},
      :exit         => 0

    option :override_runlist,
      :short        => "-o RunlistItem,RunlistItem...",
      :long         => "--override-runlist RunlistItem,RunlistItem...",
      :description  => "Replace current run list with specified items",
      :proc         => lambda { |items| items.split(',').map { |item| Chef::RunList::RunListItem.new(item) }}

    def self.print_help
      instance = new
      instance.parse_options([])
      puts instance.opt_parser
      puts
      puts footer
      puts
      exit 1
    end

    def self.setup!
      self.new.parse_opts
    end

    def parse_opts
      remainder = parse_options
      environment = remainder.first
      # We have to nuke ARGV to make sure irb's option parser never sees it.
      # otherwise, IRB complains about command line switches it doesn't recognize.
      ARGV.clear
      config[:config_file] = config_file_for_shef_mode(environment)
      config_msg = config[:config_file] || "none (standalone shef session)"
      puts "loading configuration: #{config_msg}"
      Chef::Config.from_file(config[:config_file]) if !config[:config_file].nil? && File.exists?(config[:config_file]) && File.readable?(config[:config_file])
      Chef::Config.merge!(config)
    end

    private

    def config_file_for_shef_mode(environment)
      if config[:config_file]
        config[:config_file]
      elsif environment && ENV['HOME']
        Shef.env = environment
        config_file_to_try = ::File.join(ENV['HOME'], '.chef', environment, 'shef.rb')
        unless ::File.exist?(config_file_to_try)
          puts "could not find shef config for environment #{environment} at #{config_file_to_try}"
          exit 1
        end
        config_file_to_try
      elsif ENV['HOME'] && ::File.exist?(File.join(ENV['HOME'], '.chef', 'shef.rb'))
        File.join(ENV['HOME'], '.chef', 'shef.rb')
      elsif config[:solo]
        "/etc/chef/solo.rb"
      elsif config[:client]
        "/etc/chef/client.rb"
      else
        nil
      end
    end

  end

end
